### PR TITLE
Configure Travis to build, test, and deploy a Certbot snap

### PR DESCRIPTION
**Pull request by bmw**
Despite this PR (only) being ~200 lines containing mostly code copied from another repo, there is a lot going on here. For the sake of making it both easier to review and to remember some of these things in the future by referring back to this PR, I've documented a lot of noteworthy things with section headers below. With that said, it's probably not necessary to read each section unless you're interested in that topic.

The most noteworthy thing for the reviewer is **this PR should be merged and not squashed** to preserve authorship. To merge this code, once we're happy with this PR, I'll probably open a new PR squashing any commits I make in response in review comments back into a single commit to try to keep history somewhat clean. To help prevent this PR from being accidentally squashed, I'm making this a draft PR for now.

### Git history of https://github.com/basak/certbot-snap-build

I think it is worth preserving the git history of https://github.com/basak/certbot-snap-build that this PR is based on in this repo to help us track why things were done a certain way. To do this while keeping our git history somewhat clean, I took the approach described at https://stackoverflow.com/questions/1425892/how-do-you-merge-two-git-repositories/21495718#21495718 to move all history of https://github.com/basak/certbot-snap-build into a `snap` directory. I then squashed all commits so that sequential commits from the same author are one commit. I probably could have reordered commits to try and squash things a little more, but I personally don't think it's worth the trouble. Finally, I merged this rewritten history into this branch of the Certbot repo.

The contents of the `snap` directory are identical to the current contents of https://github.com/basak/certbot-snap-build before my final commit in this PR which makes the changes to make things work in this repo.

### Travis stages

This is described in general at https://docs.travis-ci.com/user/build-stages/, but I don't think we should deploy the snap if any of our tests are failing. To accomplish this, I created a "Snap" stage that builds, tests, and deploys the snap which is only executed after a "Test" stage that contains all of our other tests. The "Snap" stage will not run until the "Test" stage completes successfully.

### snap/local

This directory is ignored by `snapcraft` which I think makes it a good place to store `snap` specific scripts like `build_and_install.sh`.

See https://bugs.launchpad.net/snapcraft/+bug/1792203 for more info.

### Why remove certbot-compatibility-test from apacheconftest toxenvs?

Because it's not used. In theory, it could go in its own PR, but it'll create merge conflicts with this one so I'd personally prefer to include this simple change in this PR as well.

### Checklist for landing this PR

- [x] Squash all of my commits into one commit
- [x] Update the release instructions to have to move the snap to the beta channel
- [x] Shut down Robie's nightly builds probably by updating his repo to say that the code has moved here and deleting everything

### CODE REVIEW

**Comment by bmw** [comment]
When this PR lands, we should remove most of the code from the `master` branch of https://github.com/basak/certbot-snap-build and have it have a README pointing here.

File: .travis.yml [hunk 1]
```diff
@@ -36,7 +42,8 @@ extended-test-suite: &extended-test-suite
 matrix:
   include:
     # Main test suite
-    - python: "2.7"
+    - stage: "Test"
```
Comment (adferrand) [inline .travis.yml:45]: As it also indicated in the example code here: https://docs.travis-ci.com/user/build-stages/#how-to-define-build-stages, I think it would be useful to remind the reader that the `Test` stage will be applied to each next matrix object until another stage is explicitly set.

File: .travis.yml [hunk 2]
```diff
@@ -11,17 +11,23 @@ before_script:
   # Use Travis retry feature for farm tests since they are flaky
   - 'if [[ "$TOXENV" == "travis-test-farm"* ]]; then export TRAVIS_RETRY=travis_retry; fi'
   - export TOX_TESTENV_PASSENV=TRAVIS
+  - 'if [[ "$SNAP" == true ]]; then snap/local/build_and_install.sh; fi'
```
Comment (adferrand) [inline .travis.yml:14]: It bothers me that this command is "simply" a part of the `before_script`, because:
- it is in fact the main purpose of this pipeline, and should benefit from a correct visibility in the pipeline for troubleshooting purpose
- it will run for each associated task, so one for the test task, and one for the snap task, while ideally, it should be an artifact that is reused on all tasks that needs it
- for multiarch build, this overhead will become very high
- failure to build the snap is a fail fast condition here: associated to the multiple call of snap builds described above, we will start several time the same broken build, wasting resources and time during the release and the nightly builds

In general I think this should be a dedicated stage, that would be before the `test` one. But this implies to be able to store pipeline artifacts, and Travis CI does not support that (it has only the capability to push artifacts to external storages, see https://docs.travis-ci.com/user/uploading-artifacts/). So I do not really have a solution here, but wanted to note that the situation is suboptimal, if you have an alternative in mind.

File: .travis.yml [hunk 3]
```diff
@@ -11,17 +11,23 @@ before_script:
...
+# will also trigger tests. This reduces the number of simultaneous Travis runs,
+# which speeds turnaround time on review since there is a cap of on the number
+# of simultaneous runs.
 branches:
   # When changing these branches, please ensure the documentation under
   # "Running tests in CI" is still correct.
   only:
     - master
     - /^\d+\.\d+\.x$/
+    - /^v\d+\.\d+\.\d+$/
```
[the remark this answers is not in the document]
Comment (bmw) [reply .travis.yml:30]: > Could you indicate on which use case each regex refers to?

Done in commits I'll push in a bit.

> It makes me think that, for the commit that is on a branch release and is tagged for the actual release, the Travis build will be run twice.

I thought about this too, but I couldn't come up with an approach I like as I would like to continue to have the full test suite run on commits to point release branches since it automates the process of running tests before a point release. I have ideas I could describe to you, but they all have downsides that to me outweigh this optimization of saving us 1 full Travis build every release.

Unless you have a concrete suggestion, I'd prefer to ignore this inefficiency until it actually becomes annoying at which point things like other CI systems may have more options available for us.

File: .travis.yml [hunk 4]
```diff
@@ -220,6 +227,48 @@ matrix:
...
+          # Deploy on release tags or nightly runs from any branch. We only try
+          # to deploy from the certbot/certbot repo to prevent errors if forks
+          # of this repo try to run tests.
+          all_branches: true
+          condition: -n $TRAVIS_TAG || $TRAVIS_EVENT_TYPE = cron
+          repo: certbot/certbot
+        provider: snap
+        snap: certbot_*.snap
+        channel: edge
+        skip_cleanup: true
```
[the remark this answers is not in the document]
Comment (bmw) [reply .travis.yml:270]: I coped it from https://github.com/basak/certbot-snap-build/blob/73c8be3cb7cfae3a7409b50ecda983c9ec501ded/.travis.yml#L22 but the reason for it is described at https://docs.travis-ci.com/user/deployment#uploading-files-and-skip_cleanup. I added a link to that in the config file here.

File: .travis.yml [hunk 5]
```diff
@@ -220,6 +227,48 @@ matrix:
...
+        'on':
+          # Deploy on release tags or nightly runs from any branch. We only try
+          # to deploy from the certbot/certbot repo to prevent errors if forks
+          # of this repo try to run tests.
+          all_branches: true
+          condition: -n $TRAVIS_TAG || $TRAVIS_EVENT_TYPE = cron
+          repo: certbot/certbot
+        provider: snap
+        snap: certbot_*.snap
+        channel: edge
```
[the remark this answers is not in the document]
Comment (bmw) [reply .travis.yml:269]: My plan for now was to continue to restrict the credentials in Travis to the edge channel and then we can manually click a button to move the snap built on the tagged commit from the edge channel to the beta channel.

We can talk about maybe changing this in the future, but it's what I'd like to do for now.

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    source-subdir: certbot
+    constraints: [$SNAPCRAFT_PART_SRC/constraints.txt]
+    python-version: python3
+    after: [acme]
+    override-pull: |
+        snapcraftctl pull
+        snapcraftctl set-version `cd $SNAPCRAFT_PART_SRC && git describe|sed s/^v//`
+    # Workaround for lack of site-packages leading to empty sitecustomize.py
+    stage:
+      - -usr/lib/python3.6/sitecustomize.py
```
[the remark this answers is not in the document]
Comment (bmw) [reply snap/snapcraft.yaml:64]: No idea. It comes from Robie's repo though and it doesn't seem to break anything so I'd prefer to leave it as is and maybe look into removing it later.

[the remark this answers is not in the document]
Comment (adferrand) [reply snap/snapcraft.yaml:64]: Ok. And same comment as below about the necessity to take care of the python version value for future maintenances.

File: snap/snapcraft.yaml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    plugin: python
+    source: .
+    source-subdir: certbot-apache
+    constraints: [$SNAPCRAFT_PART_SRC/constraints.txt]
+    python-version: python3
+    after: [python-augeas, certbot]
+    stage-packages: [libaugeas0]
+    stage:
+      # Prefer cffi
+      - -lib/python3.6/site-packages/augeas.py
```
[the remark this answers is not in the document]
Comment (bmw) [reply snap/snapcraft.yaml:75]: I think I may be able to wrangle python-augeas to remove the need for this altogether (in a separate PR), but I personally prefer to not rely on inline comments to remember to do things so I mentioned this at https://github.com/certbot/certbot/issues/7854#issuecomment-618684487.

If you'd still prefer I also add a comment here, let me know.

File: .travis.yml (hunk 2 above)
Comment (bmw) [reply .travis.yml:14]: I agree this is suboptimal and I think the best approach if we wanted to fix this would be to switch to a different CI provider (at least for the snap builds) as you mentioned in your overall review comment. We could also potentially look into using something like `s3` as shown at that link, but at that point, we're basically rolling our own version of features offered by other CI providers so I think I'd rather just switch.

If we wanted to switch, I think Azure Pipelines is really nice, but I find some of the authentication stuff like https://github.com/certbot/certbot/issues/7574 pretty frustrating for an open source project. GitHub Actions is potentially pretty promising with the downside of potentially slower builds and no tests for older versions of Windows. Circle CI is another option, however, I think we'd have to pay to get equivalent features as the first two options. That's not ideal, but we could maybe do that if once we pay, they offer the best product. I'm not sure if there are other options and I think switching requires more research/testing, but I'm personally leaning towards GitHub Actions right now.

With that said, I'm personally not that worried about this right now either. It's inelegant and inefficient, but I think it'll work, especially as an initial version of building our snaps.

File: snap/snapcraft.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (bmw) [reply snap/snapcraft.yaml:64]: I updated my comment at https://github.com/certbot/certbot/issues/7854#issuecomment-618684487 about this.

**Comment by bmw** [comment]
I didn't make all the suggested changes, but I implemented the ones I agreed with and responded to the others.

File: snap/snapcraft.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (adferrand) [reply snap/snapcraft.yaml:75]: No it is ok, your alternative is better!

File: snap/snapcraft.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (adferrand) [reply snap/snapcraft.yaml:64]: See below.

File: .travis.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (adferrand) [reply .travis.yml:278]: Ok !

File: .travis.yml (hunk 2 above)
Comment (adferrand) [reply .travis.yml:14]: I am really hesitant to approve a code that you and I consider as not great, in particular considering the high level of quality generally expected in contributions to Certbot codebase. Indeed I consider that having a strong pipeline overcomes the problems with Mattermost interconnectivity, considering that it concerns very few people while the snap build is for users. Also Azure Pipelines contributors made good improvements in GitHub PR checks in the scope of open source contributions, as anyone can check the pipelines executed on Azure without a Microsoft account.

That is why I am still in favor in finding a better fitted tool, and Azure Pipeline has my preference. I also appreciate Circle CI and use it in several projects, but it needs a paid account to scale. GitHub Actions from my latest checks seems still quite beta, and is for now a subset (in term of functionalities and performances) of Azure Pipelines.

Anyway moving to another CI tool would be a long term solution and time consuming. Right now Travis and Azure Pipelines are already integrated in Certbot, so are well positionned to limit technological risks.

If you want, I can open on my own a draft PR for Azure Pipelines, and we can compare the two approaches.

File: .travis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (adferrand) [reply .travis.yml:274]: Ok I see. Good for me. I am still hoping that we will do a real continuous delivery in the future.

[the remark this answers is not in the document]
Comment (bmw) [reply .travis.yml:274]: Me too! Like with the Windows installer, I'd just like us to carefully think through the security tradeoffs we're making and decide what we're comfortable with. Until we've done that, I'd like there to be a manual step between any automated system and the code run by (most) of our users. (Things like the edge channel and Docker Hub which we have to trust to host our Docker images anyway are exceptions at least in my mind.)

File: .travis.yml (hunk 2 above)
Comment (bmw) [reply .travis.yml:14]: I'm surprised you think that the downsides of Travis are so great here that it is worth blocking making progress on snaps to rewrite our CI system. I say rewrite our CI system because the only purpose of stages as offered by Travis and used in this PR is to prevent the snap deployment if any of our other tests are failing. An equivalent approach in any other CI system such as Azure requires migrating all of our Travis tests to that system which is a significant undertaking and probably in all cases results in some downsides like we've discussed with the different CI providers.

What really are the downsides of Travis here?

> - it is in fact the main purpose of this pipeline, and should benefit from a correct visibility in the pipeline for troubleshooting purpose

In terms of the output shown by Travis, the only thing I can think of is the snap build output is folded when it passes which seems very minor to me. In terms of understanding this file, it also seems pretty minor. I could define the build differently here, but if you really think a change is required to unblock this work, I think the best approach would simply be to add a comment saying what `SNAP` does in the matrix below so it can be clearly and immediately understood.
> - it will run for each associated task, so one for the test task, and one for the snap task, while ideally, it should be an artifact that is reused on all tasks that need it
> - for multiarch build, this overhead will become very high
> - failure to build the snap is a fail fast condition here: associated to the multiple call of snap builds described above, we will start several time the same broken build, wasting resources and time during the release and the nightly builds

Unless I'm misunderstanding something, all of these concerns are about the potential downsides caused by future, unreviewed work. In this PR there is only one Travis job that defines `SNAP`. In all other jobs, this only adds a check of an environment variable.

If when we test it the overhead for multiarch builds which will mainly run while people aren't working on Certbot is unacceptable, maybe it makes sense to block that on rewriting our CI setup, but blocking us merging Robie's code into this repo on potential unacceptable inefficiencies in that work doesn't really make sense to me.

Out of all of the concerns you mentioned, unless I'm missing something, the only one that affects the state of things now is being able to understand what Travis does here which I don't think is that hard and seems like can be more easily resolved than abandoning Travis entirely.

I am in favor of careful design decisions, but I'm also in favor of small, incremental progress where possible and the phrase "the perfect is the enemy of the good". The purpose of this PR is to set up simple snap builds by mainly just merging Robie's code into this repo with minimal changes, unburdening him from maintaining it, and making it easier for us to continue to improve it as I don't think we can/should land many more changes in his repo. I think it accomplishes that purpose pretty well without any major current downsides.

Unless you can identify some significant current problems, I'd like to move forward with this PR as is and we can discuss rewriting our CI setup in the future. You working on that could be valuable, although as usual, the amount of time I'm willing to dedicate to that myself largely depends on if I can be convinced that it is the most important thing for us to be working on with our limited resources right now.

**Comment by bmw** [comment]
I had previously put some things in comments in this PR, but to try and make things easier, I added a checklist of things I think we should do before merging this PR to the bottom of my initial comment here.

File: .travis.yml (hunk 2 above)
Comment (adferrand) [reply .travis.yml:14]: My apologies, I think there is a misunderstanding because I did not interpretate correctly your intention in this PR. I was not thinking about migrating any tests from Travis to Azure, but just to write the specific logic of building, testing and deploying snaps into Azure instead of Travis because of the various reasons described in this PR comments.

But in my mind testing snaps was about launching integration tests only (so the `integration-external` tox target you added). Similarly to what was done in Robie repo originally.

So again in my request was concerning only the small incremental change done by this PR, which was then to "just" add a continuous delivery of snaps to the edge channel after a single test, and not imply to rethink right now the all CI system.

However I see know that your intention is to wait for all tests existing in Travis to succeed before pushing the snap to edge. I understand now more the reticence you have, and the trade your are doing between accepting a suboptimal Travis pipeline for snaps build versus a major refactoring of the CI system in Certbot.

Finally it comes all up to that question: is it really needed to wait for all tests to be run before pushing the snap? In my opinion, this is not necessary, and the `integration-external` is sufficient, for the following reasons:
- several tests (basically almost all from the extended tests suite) are not concerning snaps, either because they are testing a specific Python version, or testing Certbot installed bare on specific systems
- remaining tests that are testing the internal logic of Certbot are already executed continuously on master, either through PR integrations, or in nightly builds, giving us time to fix things before release time
- outcome of snap build would be for the edge channel, which I think is acceptable to do with limited blocking tests, in particular considering that tests that would fail at the time the snap is pushed and that we would not have the opportunity to see failing before would be from the extended test suite, which is not concerning the snap is I said before
- pushing the snap to beta is a manual action, giving the releaser the opportunity to review all outcomes of the release branch (including the extended test suite from Travis), and not give his sign off if there is any doubt

I sounds to me a quite agile workflow, with shorter multiple pipelines for different purposes even allowing "live" testing thanks to the edge channel if weird things are detected in the tests.

So it is now a trade-off between a suboptimal Travis pipeline for snaps versus not having the extended test suite blocking the push to edge channel because it is done in a dedicated pipeline in Travis. If you still think we really need to have all tests run before pushing to edge channel, I am totally agree with you that it does not worth to rethink the CI system right now. Otherwise, well I think it really makes sense to have the specific snap build/integration-test/deploy in Azure Pipeline.

What do you think of that? Sorry again for the verbosity of the discussion because we were not on the same page, because of a misunderstanding on my side.

Comment (bmw) [reply .travis.yml:14]: I do not think it is strictly necessary to block deploying the snap to edge on all tests passing, however, I think it's the behavior I'd ideally like to have as I can think of some (unlikely) scenarios where there are problems affecting the snap not caught by our tests on it. Maybe we found an incompatibility between Certbot and boulder. Maybe our unit tests or mypy found a problem in a code path not tested by the integration tests run on the snap. This is possible even if these tests are run on PRs because we do not require PRs to be up-to-date with `master` before being merged. Basically, if we've found unexpected problems in our code and we're unsure if they affect the snap, why deploy it?

The main downside I can think of in blocking deployments in this way limiting the discussion to the state of things in this PR is our full test suite [takes ~8 minutes longer](https://travis-ci.org/github/bmw/letsencrypt/jobs/678394225) because it's not run in parallel. I don't think this really matters at all for nightly or release builds, but it is kind of annoying for `test-` branches.

(The way things should ideally work in my mind is everything is test and built in parallel with deployments only happening on nightly/release builds after everything else has passed. This requires not only having better support for pipelines/stages than Travis has, but everything being in the same CI provider (or in different CI providers with very good integration with each other).)

Stepping back from that though, again my goal here was simply to merge Robie's code into this repo. I mainly used Travis because Robie used Travis so it made the migration slightly easier. I think almost all of your concerns about Travis are about interactions with future work built on this and are not caused by this PR alone. As far as I can tell, what we have here works just fine with no significant downsides, takes the code from Robie with minimal changes, and sets up snap builds nightly and for future releases.

If that's correct, I'd still personally like to move forward with this PR and save potential future problems for potential future work. If you want, I can remove stages entirely with this PR which eliminates the extra 8 minutes spent testing at the cost of potentially breaking things for people caused by problems caught by our other tests. I'm also happy to add a comment explaining what `SNAP` does when we set it for the snap job.

File: .travis.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (bmw) [reply .travis.yml]: This is mainly relevant for our discussion at https://github.com/certbot/certbot/pull/7932#discussion_r414089205, but I wanted to note that we currently have this problem in Azure too where [advanced.yml](https://github.com/certbot/certbot/blob/4cce3458f3e6a38fbb3f39613aca1c9e1ebd86ca/.azure-pipelines/advanced.yml) is run on pushes to `*.x` branches and [release.yml](https://github.com/certbot/certbot/blob/4cce3458f3e6a38fbb3f39613aca1c9e1ebd86ca/.azure-pipelines/release.yml) is run on release tags.

File: .travis.yml (hunk 2 above)
Comment (adferrand) [reply .travis.yml:14]: No, for 8 min it does not worth it. We will talk about it in the future when the overhead will be more than 20 mins.

I close the discussion, for the sole purpose of what this PR wants to achieve, it does it.

**Comment by bmw** [comment]
Closing in favor of https://github.com/certbot/certbot/pull/7948.
